moesif-servlet/src/main/java/com/moesif/servlet/MoesifFilter.java

```diff
@@ -427,6 +444,143 @@ private EventResponseModel getEventResponseModel(LoggingHttpServletResponseWrapp
     return eventResponseBuilder.build();
   }
 
+  /***
+   * Method to initialize the batch event processor and create jobs for automatic update of
+   * app config and send events in batches.
+   * Batch event processor maintains batch queue and a taskRunner method to
+   * run scheduled task periodically to send events in batches.
+   */
+  public void initBatchProcessorAndStartJobs() {
+
+    // Create event batch processor for queueing and batching the events.
+    this.batchProcessor = new BatchProcessor(this.moesifApi, this.config, this.debug);
+
+    // Initialize the timer tasks - Create scheduled jobs
+    this.scheduleAppConfigJob();
+    this.scheduleBatchEventsJob();
+  }
+
+  /***
+   * Method to stop the scheduled jobs and drain the batch queue.
+   * Batch event processor sends the leftover events present in queue
+   * before stopping the scheduled jobs.
+   */
+  public void drainQueueAndStopJobs() {
+
+    // Cleanup/transfer the leftover queue events, if any before destroying the filter.
+    this.batchProcessor.run();  // it's ok to run in main thread on destroy.
+
+    // Stop the scheduled jobs
+    try {
+      if (debug) {
+        logger.info("Stopping scheduled jobs.");
+      }
+      this.resetJobTimer(this.updateConfigTimer);
+      this.resetJobTimer(this.sendBatchEventTimer);
+    } catch (Exception e) {
+      // ignore the error.
+    }
+  }
+
+  /**
+   * Method to create scheduled job for updating config periodically.
+   */
+  private void scheduleAppConfigJob() {
+    // Make sure there is none before creating the timer
+    this.resetJobTimer(this.updateConfigTimer);
+
+    this.updateConfigTimer = new Timer("moesif_update_config_job");
+    updateConfigTimer.schedule(new TimerTask() {
+      @Override
+      public void run() {
+        // get and update config.
+        getAndUpdateAppConfig();
+      }
+    }, 0, (long) this.batchProcessor.getUpdateConfigTimeInMin() * 60 * 1000);
+  }
+
+  /**
+   * Method to create scheduled job for sending batch events periodically.
+   */
+  private void scheduleBatchEventsJob() {
+    // Make sure there is none before creating the timer
+    this.resetJobTimer(this.sendBatchEventTimer);
+
+    this.sendBatchEventTimer = new Timer("moesif_events_batch_job");
+    sendBatchEventTimer.schedule(
+            this.batchProcessor,
+            0,
+            (long) this.batchProcessor.getBatchMaxTimeInSec() * 1000
+    );
+  }
+
+  /**
+   * Method to reset scheduled job timer.
+   * @param timer : Timer
+   */
+  private void resetJobTimer(Timer timer) {
+    if (timer != null) {
+      timer.cancel();
+      timer.purge();
+      timer = null;
+    }
+  }
+
+  private void rescheduleSendEventsJobIfNeeded() {
+
+    // Check if batchJob is already running then return
+    if (this.batchProcessor.isJobRunning()) {
```

Comment:
Great to see not spawning too many threads but also verifying events not stuck forever with auto reschedule